README.md
# The Kodi _Sandmann_ Addon

This is the repository for the _Kodi Sandmann Plugin_.
[Kodi](https://kodi.tv) is a famous media center software.
This plugin enables you to access video content from
[sandmann.de](http://www.sandmann.de) from within Kodi.

## Version 2 (wip)

I'm currently working on version 2 of this plugin.
Feel free to get in touch, ask questions or discuss in the [issues section](https://github.com/sorax/plugin.video.sandmann/issues)

## Bugs

If you find a bug, please submit it to GitHub's
[bug tracker](https://github.com/sorax/plugin.video.sandmann/issues)

## Contributors

- [sorax](https://github.com/sorax), [handfrog](https://github.com/handfrog)
- Special thanks to [chefmagier](https://github.com/chefmagier), [yafp](https://github.com/yafp)


changelog.txt
version 2.1.0:
 * add setting to switch between different versions (rbb/mdr/kika)
version 2.0.1:
 * update contributers
version 2.0.0:
 * complete rewrite for kodi 19 (aka matrix)
version 1.2.2:
 * Kodi 19 compatibility (respond to issue #4)
version 1.2.1:
 * fix a bug where the update-interval was initiated multiple times
version 1.2.0:
 * list all episodes
 * add switch for sign language
 * fix a bug with the "&" in episode titles
version 1.1.0:
 * add reload functionality (can be configured in the addon-settings)
version 1.0.3:
 * fix images not loading (removed query param)
version 1.0.2:
 * change json import
 version 1.0.1:
 * python3 compatibility
version 1.0.0:
 * initial version


libs/episodes.py
# -*- coding: utf-8 -*-
# Copyright 2022 sorax
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import json

try:
    from urllib.request import urlopen
except ImportError:
    from urllib2 import urlopen


def getJsonFromUrl(url):
    document = urlopen(url).read()
    return json.loads(document)


def getEpisodes(episodes_url, quality):
    episodes_json = getJsonFromUrl(episodes_url)
    episodes_list = episodes_json["teasers"]

    item_list = []
    for episode in episodes_list:
        item_list.append(getEpisodeData(episode, quality))

    return item_list


def getEpisodeData(data, quality):
    episode_url = data["links"]["target"]["href"]
    episode_details = getJsonFromUrl(episode_url)
    streams_data = episode_details["widgets"][0]["mediaCollection"]["embedded"]["_mediaArray"][0]
    mediaStreamArray = streams_data["_mediaStreamArray"]

    streams = {}
    for stream in mediaStreamArray:
        streams[stream["_quality"]] = stream["_stream"]

    return {
        "desc": getDescription(data),
        "duration": data["duration"],
        "fanart": getImage(data, 1920),
        "stream": getStream(streams, quality),
        "thumb": getImage(data, 640),
        "title": data["shortTitle"],
        "dgs": getDgs(data["shortTitle"])
    }


def getDescription(content):
    return content["images"]["aspect16x9"]["alt"].split(" (Quelle:")[0]


def getImage(content, width):
    return content["images"]["aspect16x9"]["src"].replace("{width}", str(width))


def getStream(streams, quality):
    index = quality - 1
    if index in streams:
        return streams[index]
    else:
        return streams["auto"]


def getDgs(title):
    return "- mit Gebärdensprache" in title


libs/sandmann.py
# -*- coding: utf-8 -*-
# Copyright 2022 sorax
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

# import xbmc
import xbmcaddon
import xbmcgui
import xbmcplugin

import sys

from libs.episodes import getEpisodes
from threading import Timer

# -- Addon --
addon = xbmcaddon.Addon()
addon_handle = int(sys.argv[1])
addon_name = addon.getAddonInfo("name")
addon_icon = addon.getAddonInfo("icon")

base_path = sys.argv[0]

# -- Constants --
episodes_url = "https://appdata.ardmediathek.de/appdata/servlet/tv/Sendung?documentId=6503982&json"

# -- Settings --
dgs = addon.getSettingInt("dgs2")
interval = addon.getSettingInt("interval2")
quality = addon.getSettingInt("quality2")
update = addon.getSettingInt("update2")

# -- Variables --
# refresh_timer = None


def sandmann():
    episodes_list = getEpisodes(episodes_url, quality)

    item_list = []
    for episode in episodes_list:
        if dgs == 0 and episode["dgs"] == False:
            item_list.append((episode["stream"], getListItem(episode), False))
        if dgs == 1:
            item_list.append((episode["stream"], getListItem(episode), False))
        if dgs == 2 and episode["dgs"] == True:
            item_list.append((episode["stream"], getListItem(episode), False))

    xbmcplugin.addDirectoryItems(addon_handle, item_list, len(item_list))

    if update == 1:
        li_refresh = xbmcgui.ListItem(label=addon.getLocalizedString(30020))
        xbmcplugin.addDirectoryItem(addon_handle, base_path, li_refresh, True)

    xbmcplugin.endOfDirectory(addon_handle)


def getListItem(item):
    li = xbmcgui.ListItem(label=item["title"])
    li.setArt({
        "thumb": item["thumb"],
        "fanart": item["fanart"]
    })
    li.setInfo(
        type="video",
        infoLabels={
            "title": item["title"],
            "plot": item["desc"],
            "duration": item["duration"]
        }
    )
    li.setProperty("IsPlayable", "true")
    return li


addon.py
# -*- coding: utf-8 -*-
# Copyright 2022 sorax
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import sys
from threading import Timer

from sandmann import getEpisodes

import xbmc
import xbmcplugin
import xbmcgui
import xbmcaddon

episodes_url = "https://appdata.ardmediathek.de/appdata/servlet/tv/Sendung?documentId=6503982&json"

addon = xbmcaddon.Addon()
# addon_name = addon.getAddonInfo("name")
# addon_icon = addon.getAddonInfo("icon")

addon_handle = int(sys.argv[1])
base_path = sys.argv[0]

quality = addon.getSettingInt("quality")
update = addon.getSettingInt("update")
interval = addon.getSettingInt("interval")
dgs = addon.getSettingBool("dgs")

refresh_timer = None


def addEpisodes():
    episodes = getEpisodes(episodes_url, quality, dgs)
    episode_list = []
    for episode in episodes:
        episode_list.append((episode["stream"], getListItem(episode), False))

    xbmcplugin.addDirectoryItems(addon_handle, episode_list, len(episode_list))


def getListItem(item):
    li = xbmcgui.ListItem(label=item["title"])
    li.setArt({
        "thumb": item["thumb"],
        "fanart": item["fanart"]
    })
    li.setInfo(
        type="video",
        infoLabels={
            "title": item["title"],
            "plot": item["desc"],
            "duration": item["duration"]
        }
    )
    li.setProperty("IsPlayable", "true")
    return li


def reload():
    xbmc.executebuiltin("Container.Refresh()")


if update == 0:
    refresh_timer
    # xbmc.executebuiltin("Notification(%s, %s, %d, %s)" %
    #                     (addon_name, "Updating...", 5000, addon_icon))
    refresh_timer = Timer(interval * 60 * 60, reload)
    refresh_timer.start()
    # refresh_timer.cancel()


if update == 1:
    li_refresh = xbmcgui.ListItem(
        label=addon.getLocalizedString(30020))
    xbmcplugin.addDirectoryItem(addon_handle, base_path, li_refresh, True)


addEpisodes()
xbmcplugin.endOfDirectory(addon_handle)
